### PR TITLE
[OSS] Balance the trainable params only

### LINKED ISSUE #261: [OSS] Partitioning can fail if very imbalanced parameters
## 🐛 Bug
As spotted in https://github.com/huggingface/transformers/issues/9156#issuecomment-747285458, some models can expose an interesting structure, with one very big tensor (which is not optimized), bigger than the sum of the other tensors. In that case, one rank will get no grads, and a couple of dependent functions don't like that

## Command
https://github.com/huggingface/transformers/issues/9156#issuecomment-747285458
seq2seq basic model run on 2 ranks: one rank gets 1 static tensor, the other one gets all the grads. Gradient clipping breaks and computations are not really optimized

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes #261, also https://github.com/huggingface/transformers/issues/9156. I suspect that this was also partially to blame with the iGPT trainings

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Yes 🙃 HuggingFace / ShardedDDP / AMP works, at least the dummy example shared

### CODE REVIEW

**Comment by stas00** [comment]
Thank you very much for fixing that, @blefaudeux

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -27,11 +26,6 @@
     _params_t = Any
 
 
-class BucketFlush(Enum):
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:30]: dead code removal

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -572,30 +575,6 @@ def _try_consume_work_handle(self) -> None:
             if work_handle.callback is not None:
                 work_handle.callback()
 
-    def _handle_trailing_buckets(self, flush_type: BucketFlush) -> None:
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:575]: same, sorry about that

File: fairscale/optim/oss.py [hunk 3]
```diff
@@ -139,7 +133,16 @@ def partition_parameters(self) -> List[List[dict]]:
                     # Add this param to rank with smallest size.
                     rank = sizes.index(min(sizes))
                     param_lists[rank].append(param)
-                    sizes[rank] += param.numel()
+
+                    # We're partitioning the optimizer state,
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:137]: this is the real change, the partitioning was not taking into account the fact that the params are trainable or not, althgouh this is what counts for the optimizer state. The test case for Huggingface was kind of pathological for that, because there was one big non trainable parameter (goes to rank 0) and then a lot of cumulatively smaller trainable parameters, which all went to rank 1. This meant that the model was effectively optimized on rank 1, hence defying the whole sharding purpose

**Comment by blefaudeux** [comment]
ping reviews, if you don't mind, I would love master to work for HuggingFace

File: tests/optim/test_oss.py [hunk 4]
```diff
@@ -181,6 +181,11 @@ def run_test_add_param_group(rank, world_size, tempfile_name):
     params = []
     for size in [4, 5, 2, 6, 4]:
         params.append(torch.rand(size, 1))
+
+    # Make sure that the params are trainable, enforces size-based partitioning
+    for p in params:
```
Comment (min-xu-ai) [inline tests/optim/test_oss.py:186]: need to add a test case where some params are not trainable too?

File: tests/optim/test_oss.py [hunk 5]
```diff
@@ -181,6 +181,11 @@ def run_test_add_param_group(rank, world_size, tempfile_name):
...
+        p.requires_grad = True
+
     o = optim.OSS(params, lr=0.1)
     assert len(o.param_groups) == 1
     o.add_param_group({"params": [torch.rand(3, 1)]})
```
Comment (blefaudeux) [inline tests/optim/test_oss.py:191]: @min-xu-ai this parameter is non trainable actually, so there's a mix of both

**Comment by blefaudeux** [comment]
The CI issue seems unrelated, pipe benchmark and host with an old cuda, I'm missing some context but could have a look late. cc @msbaines just in case

**Review by min-xu-ai — APPROVED** [review]
Really nice. Thanks for adding comments and tests.

**Comment by stas00** [comment]
@blefaudeux, when you feel all the important changes have been merged - please ping me so that I could re-test with transformers, and then it'd be great to make a new fairscale release and then we can announce it as working with transformers.

I have the brief doc ready to merge when the above has come to satisfaction https://github.com/huggingface/transformers/pull/9208 (in case you'd like to add anything there please don't hesitate to suggest) and then we can make an announcement that transformers has Sharded ZeRO features integrated. Yay!

**Comment by blefaudeux** [comment]
> @blefaudeux, when you feel all the important changes have been merged into master - please ping me so that I could re-test with transformers, and then it'd be great to make a new fairscale release on pypi and then we can announce it as working with transformers.
> 
> I have a brief doc ready to merge when the above has come to satisfaction [huggingface/transformers#9208](https://github.com/huggingface/transformers/pull/9208) (in case you'd like to add anything there please don't hesitate to suggest) and then we can make an announcement that transformers has Sharded ZeRO features from fairscale integrated. Yay!

Hi @stas00, should be good to go ! The current issue with CircleCI is unrelated, problem with the default torch install via pip being incompatible with the provided CUDA on these machines, else the two PRs required for HuggingFace (which were related to genuine issues) are landed. Please keep me posted if you encounter any issues, and thanks for the great work around that !

**Comment by stas00** [comment]
That's great!

I rebuilt and retested and everything looks good.

Would it be possible to make a new release on pypi first and then we are good to announce.
